File: src/improvement/website_metrics.py
"""Website metrics auto-update — keep site stats current.

Fetches metrics from GitHub API, test suites, and feature counts.
Provides data for injection into Astro site templates.
"""

from __future__ import annotations

import json
import os
import re
import subprocess
from dataclasses import dataclass


@dataclass
class ProjectMetrics:
    test_count: int = 0
    coverage: float = 0.0
    tool_count: int = 0
    mode_count: int = 0
    provider_count: int = 0
    github_stars: int = 0


def count_tests(project_dir: str, test_command: list[str] | None = None) -> int:
    """Count tests by running the test suite with collection only."""
    if test_command is None:
        test_command = ["python3", "-m", "pytest", "--collect-only", "-q"]

    try:
        result = subprocess.run(
            test_command,
            capture_output=True,
            timeout=60,
            cwd=project_dir,
        )
        output = result.stdout.decode()
        # Look for pytest output like "314 tests collected"
        match = re.search(r"(\d+)\s+test", output)
        if match:
            return int(match.group(1))
    except (subprocess.TimeoutExpired, FileNotFoundError):
        pass
    return 0


def count_mcp_tools(server_file: str) -> int:
    """Count @mcp.tool() decorators in an MCP server file."""
    try:
        with open(server_file) as f:
            content = f.read()
        return content.count("@mcp.tool()")
    except (FileNotFoundError, OSError):
        return 0


def count_modes(modes_dir: str) -> int:
    """Count mode files in a modes directory."""
    try:
        return len([f for f in os.listdir(modes_dir) if f.endswith(".md")])
    except (FileNotFoundError, OSError):
        return 0


def gather_metrics(
    project_dir: str,
    server_file: str | None = None,
    modes_dir: str | None = None,
    test_command: list[str] | None = None,
) -> ProjectMetrics:
    """Gather all metrics for a project."""
    metrics = ProjectMetrics()
    metrics.test_count = count_tests(project_dir, test_command)
    if server_file:
        metrics.tool_count = count_mcp_tools(server_file)
    if modes_dir:
        metrics.mode_count = count_modes(modes_dir)
    return metrics


def format_metrics_json(metrics: ProjectMetrics) -> str:
    """Format metrics as JSON for site injection."""
    return json.dumps({
        "test_count": metrics.test_count,
        "coverage": metrics.coverage,
        "tool_count": metrics.tool_count,
        "mode_count": metrics.mode_count,
        "provider_count": metrics.provider_count,
        "github_stars": metrics.github_stars,
    }, indent=2)

File: rust/src/improvement/website_metrics.rs
//! Website metrics — count tests, MCP tools, modes for site stats.

use regex::Regex;
use serde::{Deserialize, Serialize};

/// Aggregated project metrics for website display.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ProjectMetrics {
    #[serde(default)]
    pub test_count: usize,
    #[serde(default)]
    pub coverage: f64,
    #[serde(default)]
    pub tool_count: usize,
    #[serde(default)]
    pub mode_count: usize,
    #[serde(default)]
    pub provider_count: usize,
    #[serde(default)]
    pub github_stars: usize,
}

/// Count `@mcp.tool()` decorators in a server file's content.
pub fn count_mcp_tools(content: &str) -> usize {
    content.matches("@mcp.tool()").count()
}

/// Count mode files (.md) in a list of filenames.
pub fn count_modes(filenames: &[&str]) -> usize {
    filenames.iter().filter(|f| f.ends_with(".md")).count()
}

/// Parse pytest output to extract test count.
pub fn parse_test_count(output: &str) -> usize {
    let re = Regex::new(r"(\d+)\s+test").unwrap();
    if let Some(caps) = re.captures(output) {
        caps[1].parse().unwrap_or(0)
    } else {
        0
    }
}

/// Format metrics as JSON string.
pub fn format_metrics_json(metrics: &ProjectMetrics) -> String {
    serde_json::to_string_pretty(metrics).unwrap_or_else(|_| "{}".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_count_mcp_tools() {
        let content = "@mcp.tool()\ndef foo():\n    pass\n\n@mcp.tool()\ndef bar():\n    pass\n";
        assert_eq!(count_mcp_tools(content), 2);
    }

    #[test]
    fn test_count_mcp_tools_none() {
        assert_eq!(count_mcp_tools("no tools here"), 0);
    }

    #[test]
    fn test_count_modes() {
        let files = vec!["mode1.md", "mode2.md", "readme.txt", "config.json"];
        assert_eq!(count_modes(&files), 2);
    }

    #[test]
    fn test_parse_test_count() {
        assert_eq!(parse_test_count("314 tests collected"), 314);
        assert_eq!(parse_test_count("no tests"), 0);
    }

    #[test]
    fn test_format_metrics_json() {
        let metrics = ProjectMetrics {
            test_count: 100,
            coverage: 99.5,
            ..Default::default()
        };
        let json = format_metrics_json(&metrics);
        assert!(json.contains("\"test_count\": 100"));
    }
}
